x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RollupIDUpgradeIT.java

```diff
@@ -0,0 +1,284 @@
+        }
+
+        if (CLUSTER_TYPE == ClusterType.OLD) {
+            String recoverQuickly = "{\"settings\": {\"index.unassigned.node_left.delayed_timeout\": \"100ms\"}}";
+
+            Request createTargetIndex = new Request("PUT", "/target");
+            createTargetIndex.setJsonEntity(recoverQuickly);
+            client().performRequest(createTargetIndex);
+
+            final Request indexRequest = new Request("POST", "/target/_doc/1");
+            indexRequest.setJsonEntity("{\"timestamp\":\"2018-01-01T00:00:01\",\"value\":123}");
+            client().performRequest(indexRequest);
+
+            // create the rollup job
+            final Request createRollupJobRequest = new Request("PUT", "/_xpack/rollup/job/rollup-id-test");
+            createRollupJobRequest.setJsonEntity("{"
+                + "\"index_pattern\":\"target\","
+                + "\"rollup_index\":\"rollup\","
+                + "\"cron\":\"*/1 * * * * ?\","
+                + "\"page_size\":100,"
+                + "\"groups\":{"
+                + "    \"date_histogram\":{"
+                + "        \"field\":\"timestamp\","
+                + "        \"interval\":\"5m\""
+                + "      },"
+                +       "\"histogram\":{"
+                + "        \"fields\": [\"value\"],"
+                + "        \"interval\":1"
+                + "      },"
+                +       "\"terms\":{"
+                + "        \"fields\": [\"value\"]"
+                + "      }"
+                + "},"
+                + "\"metrics\":["
+                + "    {\"field\":\"value\",\"metrics\":[\"min\",\"max\",\"sum\"]}"
+                + "]"
+                + "}");
+
+            Map<String, Object> createRollupJobResponse = toMap(client().performRequest(createRollupJobRequest));
+            assertThat(createRollupJobResponse.get("acknowledged"), equalTo(Boolean.TRUE));
+
+            Request updateSettings = new Request("PUT", "/rollup/_settings");
+            updateSettings.setJsonEntity(recoverQuickly);
+            client().performRequest(updateSettings);
+
+            // start the rollup job
+            final Request startRollupJobRequest = new Request("POST", "_xpack/rollup/job/rollup-id-test/_start");
+            Map<String, Object> startRollupJobResponse = toMap(client().performRequest(startRollupJobRequest));
+            assertThat(startRollupJobResponse.get("started"), equalTo(Boolean.TRUE));
+
+            assertRollUpJob("rollup-id-test");
+            List<String> ids = getSearchResults(1);
+            assertThat(ids, containsInAnyOrder("3310683722"));
+
+        }
+
+        if (CLUSTER_TYPE == ClusterType.MIXED && Booleans.parseBoolean(System.getProperty("tests.first_round"))) {
+            final Request indexRequest = new Request("POST", "/target/_doc/2");
+            indexRequest.setJsonEntity("{\"timestamp\":\"2018-01-02T00:00:01\",\"value\":345}");
```

Comment:
kind of a nit, but when I've done this kind of thing, I always regret it when I use hardcoded timestamps. Can we refactor to use a base timestamp, and then use date math to add a day and such? that would make it easier to randomize base timestamp, and more clear what the test is doing.

Reply:
We can... but it will complicate the test and may make it flakier.

The IDs that are generated in rollup docs are essentially a concatenation of the composite agg keys, which are then hashed (either CRC32 in old, or Murmur3 in new).  The composite keys will be values like the `"2018-01-02T00:00:01"` and `345`.

So by hardcoding the timestamps we know what IDs are going to be generated at each phase of the rolling upgrade and can test for them explicitly.

If we randomize, we'll also have to run the values through the ID generator to create the hashes.  I always get worried when I have to use part of the thing under test to verify the test itself.  E.g. if we break the ID generator in the future, the "broken" IDs would then change in the test and we might not notice.

We do have some unit tests that check the ID itself (`IndexerUtilsTests#testKeyOrderingOldID()`, `IndexerUtilsTests#testKeyOrderingNewID()`, `IndexerUtilsTests#testKeyOrderingNewIDLong()`) so maybe it's not an issue... but it does cause me a bit of concern.

WDYT?

In either case, I can document what this test is doing so it's more clear that the values being indexed are directly affecting the IDs which are being verified.

Reply:
that seems fine. fwiw, I am not saying we should necessarily randomize, but that for readability and maintainability, it's clearer to have (forgive the pseudocode)

```
DateTime baseDate = new DateTime("2010-01-01")
indexRequest.setJsonEntity("timestamp"+baseDate.plusDays(1).toString());
....
indexRequest.setJsonEntity("timestamp"+baseDate.plusDays(2).toString());
...
```

again this is kind of a nit but I've found as test scope grows, that hardcoded dates that depend on each other throughout the tests are hard to manage, and mutating a base date makes refactoring things easier and clearer to our future selves what the test is doing.

Reply:
Roger, parameterizing but not-randomizing seems fine with me.  Will make the change :)